<a href="https://colab.research.google.com/github/MarioROT/IHLT-MAI/blob/main/Session3_MarioRosas_AlamLopez.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>



```
```

# Lab session 3 (Morphology) - IHLT

**Students:**
- Mario Rosas
- Alam Lopez

**Lab Professor:** Salvador Medina Herrera

## Paraphrases Template

In [1]:
%%shell
git clone https://github.com/mariorot/IHLT-MAI.git
cd 'IHLT-MAI'
mv 'complementary_material' /content/
mv scripts /content/

Cloning into 'IHLT-MAI'...
remote: Enumerating objects: 86, done.
remote: Counting objects: 100% (86/86), done.
remote: Compressing objects: 100% (73/73), done.
remote: Total 86 (delta 39), reused 43 (delta 12), pack-reused 0
Receiving objects: 100% (86/86), 139.41 KiB | 1.79 MiB/s, done.
Resolving deltas: 100% (39/39), done.


In [1]:
from scripts.compute_metrics import ComputeMetrics
from scripts.text_preprocessing import TextPreprocessing
import pandas as pd
from scipy.stats import pearsonr
import nltk
nltk.download('averaged_perceptron_tagger')
nltk.download('wordnet')
nltk.download('omw-1.4')
import spacy
nlp = spacy.load("en_core_web_sm")


[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Package punkt is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!
[nltk_data] Downloading package averaged_perceptron_tagger to
[nltk_data]     /root/nltk_data...
[nltk_data]   Package averaged_perceptron_tagger is already up-to-
[nltk_data]       date!
[nltk_data] Downloading package wordnet to /root/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!
[nltk_data] Downloading package omw-1.4 to /root/nltk_data...
[nltk_data]   Package omw-1.4 is already up-to-date!


In [2]:
dt = pd.read_csv('complementary_material/test-gold/STS.input.SMTeuroparl.txt',sep='\t',header=None)
dt['gs'] = pd.read_csv('complementary_material/test-gold/STS.gs.SMTeuroparl.txt',sep='\t',header=None)

## TODO
1. Read all pairs of sentences of the SMTeuroparl files of test set within the evaluation framework of the project. Compute their similarities by considering *lemmas* and *Jaccard distance*. <br>

2. Compare the results with those in session 2 (document structure) in which words were considered. : <br>

3. Compare the results with gold standard by giving the pearson correlation between them. <br>

4. Questions (justify the answers): <br>

      Which is better: words or lemmas?

      Do you think that could perform better for any pair of texts?
```
from scipy.stats import pearsonr
pearsonr(dt['gs'], dt['jaccard'])[0]
```

The code for the tokenization is at: https://github.com/MarioROT/IHLT-MAI/blob/main/scripts/text_preprocessing.py

The code for jaccard distance computation is at: https://github.com/MarioROT/IHLT-MAI/blob/main/scripts/compute_metrics.py

In [3]:
tp = TextPreprocessing()
dt[2] = tp.tokenize_data(list(dt[0]),'nltk')
dt[3] = tp.tokenize_data(list(dt[1]),'nltk')
cm_nltk = ComputeMetrics(dt[[2,3]].to_numpy(), ['jaccard'], 1)
dt['jaccard_nltk'] = cm_nltk.do()['jaccard']

dt[4] = tp.tokenize_data(list(dt[0]),'spacy')
dt[5] = tp.tokenize_data(list(dt[1]),'spacy')

cm_spacy = ComputeMetrics(dt[[4,5]].to_numpy(), ['jaccard'], 1)
dt['jaccard_spacy'] = cm_spacy.do()['jaccard']

Computing jaccard...
Computing jaccard...


In [5]:
tp.lemmatize_data(list(dt[0]),'spacy')

[['the',
  'leader',
  'have',
  'now',
  'be',
  'give',
  'a',
  'new',
  'chance',
  'and',
  'let',
  'we',
  'hope',
  'they',
  'seize',
  'it',
  '.'],
 ['amendment',
  'no',
  '7',
  'propose',
  'certain',
  'change',
  'in',
  'the',
  'reference',
  'to',
  'paragraph',
  '.'],
 ['let',
  'I',
  'remind',
  'you',
  'that',
  'our',
  'ally',
  'include',
  'fervent',
  'supporter',
  'of',
  'this',
  'tax',
  '.'],
 ['the', 'vote', 'will', 'take', 'place', 'today', 'at', '5.30', 'p.m.'],
 ['the',
  'fisherman',
  'be',
  'inactive',
  ',',
  'tired',
  'and',
  'disappointed',
  '.'],
 ['neither',
  'be',
  'there',
  'a',
  'qualified',
  'majority',
  'within',
  'this',
  'House',
  'to',
  'revert',
  'to',
  'Article',
  '272',
  '.'],
 ['it',
  'increase',
  'the',
  'power',
  'of',
  'the',
  'big',
  'country',
  'at',
  'the',
  'expense',
  'of',
  'the',
  'small',
  'country',
  '.'],
 ['the',
  'fisherman',
  'be',
  'inactive',
  ',',
  'tired',
  'and',
  '

In [ ]:
dt[['gs','jaccard_nltk','jaccard_spacy']]

gs  jaccard_nltk  jaccard_spacy
0    4.500      0.653846       0.653846
1    5.000      0.214286       0.214286
2    4.250      0.608696       0.608696
3    4.500      0.454545       0.400000
4    5.000      0.000000       0.000000
..     ...           ...            ...
454  5.000      0.450000       0.450000
455  4.750      0.642857       0.714286
456  5.000      0.600000       0.600000
457  4.000      0.681818       0.681818
458  3.833      0.500000       0.500000

[459 rows x 3 columns]

In [ ]:
p_nltk=pearsonr(dt['gs'], 1-dt['jaccard_nltk'])[0] # Calculating the pearson correlation between GS results and 1-Jaccard calculated dats
p_spacy=pearsonr(dt['gs'], 1-dt['jaccard_spacy'])[0]
print(f'NLTK {p_nltk} --- spaCy {p_spacy}')

NLTK 0.45049771693186846 --- spaCy 0.4608504322255335


In [ ]:
words = ['women', 'played', 'with', 'small', 'children', 'happily']
nltk.pos_tag(words)

[('women', 'NNS'),
 ('played', 'VBD'),
 ('with', 'IN'),
 ('small', 'JJ'),
 ('children', 'NNS'),
 ('happily', 'RB')]

In [ ]:

def lemmatize(p):
  d = {'NN': 'n', 'NNS': 'n',
       'JJ': 'a', 'JJR': 'a', 'JJS': 'a',
       'VB': 'v', 'VBD': 'v', 'VBG': 'v', 'VBN': 'v', 'VBP': 'v', 'VBZ': 'v',
       'RB': 'r', 'RBR': 'r', 'RBS': 'r'}
  if p[1] in d:
    return wnl.lemmatize(p[0], pos=d[p[1]])
  return p[0]

# Conclusion
The use of different tokenizers (NLTK and spaCy) can impact the results when comparing sentence pairs, as they may perform differently depending on the input text data. Jaccard similarity proved to be an effective metric for comparing the similarity of sentence pairs after word tokenization. With the inversion of this metric we got a distance to measure the dissimilarity between sets of tokens, making it suitable for the task we were trying to solve. The Pearson correlation coefficient provided us a quantitative measure of how well our results align with the gold-standard, i.e. effectiveness.

The results indicate that spaCy slightly outperformed NLTK in the task of sentence pair comparison using Jaccard distance.The Pearson correlation coefficient values obtained for **NLTK (0.4505)** and **spaCy (0.4609)** suggest a moderate positive correlation between the computed distances and the benchmark (gold-standard) distances.

Our best result could be possibly improved by experimenting with different tokenization strategies or exploring other NLP techniques to enhance the performance.


